src/main/java/com/nutomic/syncthingandroid/fragments/DeviceFragment.java

```diff
@@ -0,0 +1,405 @@
+
+    private TextView mSyncthingVersionView;
+
+    private View mCompressionContainer;
+
+    private boolean mIsCreateMode;
+
+    private DialogInterface.OnClickListener mCompressionEntrySelectedListener = new DialogInterface.OnClickListener() {
+        @Override
+        public void onClick(DialogInterface dialog, int which) {
+            dialog.dismiss();
+
+            Compression compression = Compression.fromIndex(which);
+            // Don't pop the restart dialog unless the value is actually different.
+            if (compression != Compression.fromValue(getActivity(), mDevice.compression)) {
+                mDevice.compression = compression.getValue(getActivity());
+                mCompressionValueView.setText(compression.getTitle(getActivity()));
+                updateDevice();
+            }
+        }
+    };
+
+    private TextWatcher mIdTextWatcher = new TextWatcherAdapter() {
+        @Override
+        public void afterTextChanged(Editable s) {
+            mDevice.deviceID = s.toString();
+        }
+    };
+
+    private TextWatcher mNameTextWatcher = new TextWatcherAdapter() {
+        @Override
+        public void afterTextChanged(Editable s) {
+            mDevice.name = s.toString();
+        }
+    };
+
+    private TextWatcher mAddressesTextWatcher = new TextWatcherAdapter() {
+        @Override
+        public void afterTextChanged(Editable s) {
+            mDevice.addresses = persistableAddresses(s);
+        }
+    };
+
+    private CompoundButton.OnCheckedChangeListener mIntroducerCheckedChangeListener = new CompoundButton.OnCheckedChangeListener() {
+        @Override
+        public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
+            mDevice.introducer = isChecked;
+            updateDevice();
+        }
+    };
+
+    @Override
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        SettingsActivity activity = (SettingsActivity) getActivity();
+        mIsCreateMode = activity.getIsCreate();
+        activity.registerOnServiceConnectedListener(this);
+        activity.setTitle(mIsCreateMode ? R.string.add_device : R.string.edit_device);
+        setHasOptionsMenu(true);
+
+        if (mIsCreateMode) {
+            if (savedInstanceState != null) {
+                mDevice = (RestApi.Device) savedInstanceState.getSerializable("device");
+            }
+            if (mDevice == null) {
+                initDevice();
+            }
+        }
+    }
+
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        if (mSyncthingService != null) {
+            mSyncthingService.unregisterOnApiChangeListener(this);
+        }
+    }
+
+    @Override
+    public void onStop() {
+        super.onStop();
+
+        // We don't want to update every time a TextView's character changes,
+        // so we hold off until the view stops being visible to the user.
+        updateDevice();
```

Comment:
I just tried to share a device ID. Right after selecting an app to share to, it went back to Syncthing and showed a restart dialog.

You should move this to onPause so it shows earlier, but also check if the device was edited at all.

Also, flipping the Introducer setting in an existing device creates 2 or 3 restart dialogs in a row.